tests/test_projectbuilder.py

```diff
@@ -148,26 +146,53 @@ def test_check_dependencies(mocker):
         not builder.check_dependencies('wheel')
 
 
-@pytest.mark.skipif(sys.version_info[:2] == (3, 5), reason='bug in mock')
-def test_build(mocker):
-    open_mock = mocker.mock_open(read_data=DUMMY_PYPROJECT)
+def test_build(mocker, pyproject_mock):
     mocker.patch('importlib.import_module')
-    mocker.patch('{}.open'.format(build_open_owener), open_mock)
+    mocker.patch('{}.open'.format(build_open_owner), pyproject_mock)
     mocker.patch('pep517.wrappers.Pep517HookCaller')
 
-    builder = build.ProjectBuilder('.')
+    builder = build.ProjectBuilder()
 
     builder.hook.build_sdist.side_effect = [None, Exception]
     builder.hook.build_wheel.side_effect = [None, Exception]
 
     builder.build('sdist', '.')
-    builder.hook.build_sdist.assert_called()
+    builder.hook.build_sdist.assert_called_with('.', {})
 
     builder.build('wheel', '.')
-    builder.hook.build_wheel.assert_called()
+    builder.hook.build_wheel.assert_called_with('.', {})
 
     with pytest.raises(build.BuildBackendException):
         builder.build('sdist', '.')
 
     with pytest.raises(build.BuildBackendException):
         builder.build('wheel', '.')
+
+
+def test_default_backend(mocker, empty_file_mock):
+    mocker.patch('importlib.import_module')
+    mocker.patch('{}.open'.format(build_open_owner), empty_file_mock)
+    mocker.patch('pep517.wrappers.Pep517HookCaller')
+
+    builder = build.ProjectBuilder()
+
+    assert builder._build_system == build._DEFAULT_BACKEND
+
+
+def test_missing_backend(mocker, pyproject_no_backend_mock):
+    mocker.patch('importlib.import_module')
+    mocker.patch('{}.open'.format(build_open_owner), pyproject_no_backend_mock)
+    mocker.patch('pep517.wrappers.Pep517HookCaller')
+
+    builder = build.ProjectBuilder()
+
+    assert builder._build_system == build._DEFAULT_BACKEND
```

Comment:
I am not really an expert in testing CLIs, so maybe this is the most expedient way to do this, but my general rule for tests is that I only test the public API, so anything that's making assertions about the internal private structure is usually off-limits (or I mark it with a comment or something to indicate why this is the only / best way to test it).

Put another way, if you did an otherwise safe refactoring of the `ProjectBuilder` class, you would need to change this test, but that defeats one of the purposes of tests — allowing you to refactor with confidence, because you've made assertions about the behavior that should remain invariant.

Of course, I rarely have to test CLIs, so I think the rules may be a little different there (when there's no public API available from Python *anyway*).

Reply:
> I am not really an expert in testing CLIs, so maybe this is the most expedient way to do this, but my general rule for tests is that I only test the public API, so anything that's making assertions about the internal private structure is usually off-limits (or I mark it with a comment or something to indicate why this is the only / best way to test it).

We are not testing CLI here, we are testing `ProjectBuilder`, which is public API. python-build provides a CLI but also a minimal API to build projects, which is `ProjectBuilder`.

> Put another way, if you did an otherwise safe refactoring of the `ProjectBuilder` class, you would need to change this test, but that defeats one of the purposes of tests — allowing you to refactor with confidence, because you've made assertions about the behavior that should remain invariant.

If I do a safe refactoring of `ProjectBuilder` these tests should not fail. They might break if I change `ProjectBuilder`'s public API.

Reply:
> If I do a safe refactoring of `ProjectBuilder` these tests should not fail. They might break if I change `ProjectBuilder`'s public API.

Well, if you removed or renamed `build._DEFAULT_BACKEND` or `ProjectBuilder._build_backend`, these tests would fail, but presumably those are not part of the public interface.

It seems like our public interface exposes a `hook` attribute, which would be the `PEP517HookCaller` object, and that *does* expose a `.build_backend` attribute, publicly.

I think you'd need to ease off the mocking a bit, but you can define a top-level `DEFAULT_BACKEND` in this test file, and then do:

```python
assert builder.hook.build_backend == DEFAULT_BACKEND
```

Unless it is unintentional that `builder.hook` is part of the public API, or that it is guaranteed to be a `Pep517HookCaller` object, in which case probably mocking out the default backend, calling `.build` and then using `Mock.assert_called_once` or `Mock.assert_called_once_with` would be more approriate.